File: p721m/accounts_merge.js
/**
 * @param {string[][]} accounts
 * @return {string[][]}
 */
var accountsMerge = function(accounts) {
    let len = accounts.length;
    // 1. Id each original account and create reverse map
    let emailIdMap = new Map();
    for (let i = 0; i < len; ++i) {
        let acct = accounts[i];
        for (let j = 1; j < acct.length; ++j) {
            let e = emailIdMap.get(acct[j]);
            if (e === undefined) {
                e = [i];
            } else {
                e.push(i);
            }
            emailIdMap.set(acct[j], e);
        }
    } // console.log(emailIdMap);
    
    // 2. Use DFS to merge accounts
    let merged = []; // name * Set of emails
    let visited = Array.from({length: len}, v => false);
    // Given an acccount id, fill in all the emails linked using the emailIdMap
    let recDFS = function(i, emails) {
        if (visited[i]) { return; }
        visited[i] = true;
        let acct = accounts[i];
        for (let j = 1; j < acct.length; ++j) {
            emails.add(acct[j]);
            let links = emailIdMap.get(acct[j]);
            for (let id of links) {
                recDFS(id, emails);
            }
        }
    };
    for (let i = 0; i < len; ++i) {
        let emails = new Set(); 
        recDFS(i, emails);
        if (emails.size > 0) {
            let name = [accounts[i][0]];
            let list = [];
            emails.forEach(v => { list.push(v); });
            list.sort();
            merged.push(name.concat(list));
        }
    }
    return merged;
};
// TEST
[
    [],
    [
        ["John", "johnsmith@mail.com", "john00@mail.com"],
    ],
    [
        ["John", "johnsmith@mail.com", "john00@mail.com"],
        ["John", "johnnybravo@mail.com"],
        ["John", "johnsmith@mail.com", "john_newyork@mail.com"],
        ["Mary", "mary@mail.com"]
    ],
].forEach(t => {
    process.stdout.write("Merged accounts: ->\n");
    console.log(accountsMerge(t));
});

File: p721m/accounts_merge.py
from collections import defaultdict

# UnionFind using email as id
class UnionFind:
    def __init__(self):
        self.parent: dict[str, str] = dict()

    def root(self, i: str) -> str:
        while self.parent[i] != i:
            self.parent[i] = self.parent[self.parent[i]]  # path compression
            i = self.parent[i]
        return i

    def union(self, p: str, q: str) -> None:
        if p not in self.parent:
            self.parent[p] = p
        if q not in self.parent:
            self.parent[q] = q
        self.parent[self.root(p)] = self.root(q)

    def trees(self) -> dict:
        trees = defaultdict(list)
        for email in self.parent.keys():
            trees[self.root(email)].append(email)
        return trees


class Solution:
    def accountsMerge(self, accounts: list[list[str]]) -> list[list[str]]:
        union_find = UnionFind()
        email_to_name = dict()
        for account in accounts:
            name = account[0]
            for email in account[1:]:
                email_to_name[email] = name
                union_find.union(email, account[1])

        return [
            [email_to_name[root]] + sorted(emails)
            for root, emails in union_find.trees().items()
        ]


# TESTS
for accounts, expected in [
    (
        [
            ["John", "johnsmith@mail.com", "john_newyork@mail.com"],
            ["John", "johnsmith@mail.com", "john00@mail.com"],
            ["Mary", "mary@mail.com"],
            ["John", "johnnybravo@mail.com"],
        ],
        [
            [
                "John",
                "john00@mail.com",
                "john_newyork@mail.com",
                "johnsmith@mail.com",
            ],
            ["John", "johnnybravo@mail.com"],
            ["Mary", "mary@mail.com"],
        ],
    ),
    (
        [
            ["Gabe", "Gabe0@m.co", "Gabe3@m.co", "Gabe1@m.co"],
            ["Kevin", "Kevin3@m.co", "Kevin5@m.co", "Kevin0@m.co"],
            ["Ethan", "Ethan5@m.co", "Ethan4@m.co", "Ethan0@m.co"],
            ["Hanzo", "Hanzo3@m.co", "Hanzo1@m.co", "Hanzo0@m.co"],
            ["Fern", "Fern5@m.co", "Fern1@m.co", "Fern0@m.co"],
        ],
        [
            ["Ethan", "Ethan0@m.co", "Ethan4@m.co", "Ethan5@m.co"],
            ["Gabe", "Gabe0@m.co", "Gabe1@m.co", "Gabe3@m.co"],
            ["Hanzo", "Hanzo0@m.co", "Hanzo1@m.co", "Hanzo3@m.co"],
            ["Kevin", "Kevin0@m.co", "Kevin3@m.co", "Kevin5@m.co"],
            ["Fern", "Fern0@m.co", "Fern1@m.co", "Fern5@m.co"],
        ],
    ),
]:
    sol = Solution()
    actual = sol.accountsMerge(accounts)
    print("Merged accounts ->", actual)
    assert sorted(actual) == sorted(expected)
